### PR TITLE
[NJT][flop counter] attention: if offsets are fake, use max seqlen

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132531
* #132466
* __->__ #132356

The flop counter is used by the partitioner, in which case the tensors passed in can be fake.

The flop computations for nested attention use the offsets to determine the actual amount of compute that will be done. But when the offsets are fake, we end up with unbacked symints (from `(offsets[1:] - offsets[:-1]).to_list()`). If we find that the offsets are fake or functional tensors, then use the max sequence length instead.

Repro: https://gist.github.com/davidberard98/903fb3e586edb6d1d466786e1a610eba

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D60597463](https://our.internmc.facebook.com/intern/diff/D60597463)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60597463).

File: test/test_flop_counter.py [hunk 1]
```diff
@@ -675,6 +676,55 @@ def split_tensor(x):
...
+    @unittest.skipIf(
+        not PLATFORM_SUPPORTS_FLASH_ATTENTION,
+        "Does not support all SDPA backends (pre-SM80 hardware on CUDA)",
+    )
+    def test_nested_attention_fake_tensors(self):
+        # shape_env = ShapeEnv()
+        x = torch.randn(123, 4, 16, device="cuda", dtype=torch.bfloat16)
+        offsets = torch.tensor([0, 30, 60, 90, 123], device="cuda")
+        max_seqlen = 40
+        # with FakeTensorMode(shape_env=shape_env) as fake_mode:
```
Comment (YuqingJ) [inline test/test_flop_counter.py:690]: Do you want to remove these commented out code? Or do they both work?

**Review by YuqingJ — COMMENTED** [review]
LGTM, did you also test in the internal job too?

File: test/test_flop_counter.py (hunk 1 above)
Comment (davidberard98) [reply test/test_flop_counter.py:690]: oh, thanks for catching that

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60597463).

File: torch/utils/flop_counter.py [hunk 2]
```diff
@@ -265,6 +265,18 @@ def sdpa_flop(query_shape, key_shape, value_shape, *args, out_shape=None, **kwar
...
 
+def _offsets_to_lengths(offsets, max_len):
+    """
+    If the offsets tensor is fake, then we don't know the actual lengths.
+    In that case, we can just assume the worst case; each batch has max length.
+    """
+    from torch._subclasses.fake_tensor import FakeTensor
+    from torch._subclasses.functional_tensor import FunctionalTensor
+    if not isinstance(offsets, (FakeTensor, FunctionalTensor)):
+        return (offsets[1:] - offsets[:-1]).tolist()
```
Comment (soulitzer) [inline torch/utils/flop_counter.py:276]: nit: preexisting, but could we use `diff` instead?

**Review by soulitzer — APPROVED** [review]

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60597463).

File: torch/utils/flop_counter.py (hunk 2 above)
Comment (davidberard98) [reply torch/utils/flop_counter.py]: thanks, good point!

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60597463).

**Pull request closed without merging**